blob/azureblob/azureblob.go

```diff
@@ -0,0 +1,583 @@
+// to construct a blob.Bucket.
+//
+// Open URLs
+//
+// For blob.Open URLs, azureblob registers for the protocol "azblob".
+// The URL's Host is used as the bucket name.
+//
+// Authentication Options:
+//
+// Option 1: Use the Storage Account Name and Account Key (Primary or Secondary)
+// This option requires the following environment variables to be set: AZURE_STORAGE_ACCOUNT_NAME, AZURE_STORAGE_ACCOUNT_KEY
+//
+// Option 2: Use a Shared Access Token (SASToken) for the Storage Account or Container
+// This option requires the following environment variables to be set: AZURE_STORAGE_ACCOUNT_NAME, AZURE_STORAGE_SASTOKEN
+// See documentation on Shared Access Signature: https://docs.microsoft.com/en-us/azure/storage/common/storage-dotnet-shared-access-signature-part-1#what-is-a-shared-access-signature
+//
+// The following query options are supported:
+//  - backslashEscapeStr: Optional, defaults to forwardslash. Used to set the escape character for backslashes.
+//
+// Example URL:
+//  azblob://mybucket?backslashEscapeStr=%2F
+//
+// As
+//
+// azureblob exposes the following types for As:
+//  - Bucket: *azblob.BlockBlobURL
+//  - Error: *azblob.StorageError
+//  - ListObject: *azblob.BlobItem for objects, *azblob.BlobPrefix for "directories".
+//  - ListOptions.BeforeList: Not Implemented
+//  - Reader: *azblob.ContainerURL
+//  - Attributes: *azblob.BlobGetPropertiesResponse
+//  - WriterOptions.BeforeWrite: Not Implemented
+package azureblob
+
+import (
+	"strconv"
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/google/uuid"
+
+	"gocloud.dev/blob"
+	"gocloud.dev/blob/driver"
+
+	"github.com/Azure/azure-pipeline-go/pipeline"
+	"github.com/Azure/azure-storage-blob-go/azblob"
+)
+
+type bucket struct {
+	name             string
+	credential       *azblob.SharedKeyCredential // used for SignedURL
+	urls             serviceUrls
+	pageMarkers      map[string]azblob.Marker // temporary page marker map, until azblob.Marker is an exportable type
+	defaultDelimiter string                   // for escaping backslashes
+}
+
+type serviceUrls struct {
+	serviceURL   *azblob.ServiceURL   // represents the Azure Storage Account
+	containerURL *azblob.ContainerURL // represents the Azure Storage Container
+	blockBlobURL *azblob.BlockBlobURL // represents the Azure Block Blob
+}
+
+// Settings to establish connection to Azure
+type Settings struct {
+	AccountName      string
+	AccountKey       string
+	DefaultDelimiter string
+	SASToken         string
+	Pipeline         pipeline.Pipeline
+}
+
+const (
+	// BlobPathSeparator is used to escape backslashes
+	BlobPathSeparator = "/"
+	// OSPathSeparator or backslashes must be converted to forwardslashes
+	OSPathSeparator = "\\"
+)
+
+var (
+	maxDownloadRetryRequests = 3               // download retry policy
+	defaultPageSize          = 1000            // default page size for ListPaged
```

Comment:
The Azure SDK pagesize is 5000. If retry is not set then it will attempt to download once. Any transient error will fail the download operation. defaultPageSize of 1000 was used to keep consistent with the other providers; 3 download retries seems like a safe number